src/matter/interaction/InteractionClient.ts

```diff
@@ -76,14 +76,35 @@ export class SubscriptionClient implements ProtocolHandler<MatterController> {
     }
 }
 
+export class ExchangeProvider {
+    constructor(
+        private readonly exchangeManager: ExchangeManager<MatterController>,
+        private channel: MessageChannel<MatterController>,
+        private readonly reconnectChannelFunc?: () => Promise<MessageChannel<MatterController>>,
+    ) {}
+
+    addProtocolHandlerToExchangeManager(handler: ProtocolHandler<MatterController>) {
+        this.exchangeManager.addProtocolHandler(handler);
+    }
+
+    initiateExchange(): MessageExchange<MatterController> {
+        return this.exchangeManager.initiateExchangeWithChannel(this.channel, INTERACTION_PROTOCOL_ID);
+    }
+
+    async reconnectChannel() {
+        if (this.reconnectChannelFunc === undefined) return false;
+        this.channel = await this.reconnectChannelFunc();
+        return true;
+    }
+}
+
 export class InteractionClient {
     private readonly subscriptionListeners = new Map<number, (dataReport: DataReport) => void>();
 
     constructor(
-        private readonly exchangeManager: ExchangeManager<MatterController>,
-        private readonly channel: MessageChannel<MatterController>,
+        private readonly exchangeProvider: ExchangeProvider,
     ) {
-        this.exchangeManager.addProtocolHandler(new SubscriptionClient(this.subscriptionListeners));
+        this.exchangeProvider.addProtocolHandlerToExchangeManager(new SubscriptionClient(this.subscriptionListeners));
```

Comment:
This doesn't right... We might add the SubscriptionClient multiple time on the exchangeManager.

The ExchangeManager should be a singleton for the Device, all protocol handlers should only be added during the initialization and the list not modified afterwards.

Try to restructure the code so this is added only once.
Or we should discuss this during a review session

Reply:
I do not see why we wpuld add it multiple times - and if then the issue existed before already. If you see such cases then we should address them in a separate PR.

And yes many of this will be reworked again when we do the new Hight level API and such ... lets cleanup then please

Reply:
@mfucci Ok I checked it again.

First of all it is Controller and not Device

We have currently two places where we instanciate an InteractionClient and in the past the exchangeManager was handed in and SubscriptionClient registered. So yes it is very good possible that the SubscriptionHandler was and is registered potentiella multiple times.

Because the issue existed already before I would add a Todo and address this after the move

Reply:
PS: And when we do high level API we need to adjust that also in any case (again)